packages/ember-metal/lib/chains.js

```diff
@@ -21,6 +21,90 @@ function Chains() { }
 
 Chains.prototype = Object.create(null);
 
+// a map of nodes that reference a key
+function ChainWatchers(src) {
+  this.src = src;
+  this.nodes = Object.create(null);
+}
+
+ChainWatchers.prototype = {
+  add(key, node) {
+    let nodes = this.nodes[key];
+    if (nodes === undefined) {
+      this.nodes[key] = [node];
+    } else {
+      nodes.push(node);
+    }
+  },
+
+  remove(key, node) {
+    let nodes = this.nodes[key];
+    if (nodes) {
+      for (var i = 0, l = nodes.length; i < l; i++) {
+        if (nodes[i] === node) {
+          nodes.splice(i, 1);
+          break;
+        }
+      }
+    }
+  },
+
+  has(key, node) {
+    let nodes = this.nodes[key];
+    if (nodes) {
+      for (var i = 0, l = nodes.length; i < l; i++) {
+        if (nodes[i] === node) {
+          return true;
+        }
+      }
+    }
+    return false;
+  },
+
+  revalidateAll() {
+    for (let key in this.nodes) {
```

Comment:
@stefanpenner should it just search a single array or linked list of well shaped entries?  like {key: 'foo', node: node} ?